core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceCreator.scala

```diff
@@ -0,0 +1,688 @@
+  }
+}
+
+class WorkspaceCreator(val userInfo: UserInfo,
+                       val dataSource: SlickDataSource,
+                       val samDAO: SamDAO,
+                       val gcsDAO: GoogleServicesDAO,
+                       val config: WorkspaceServiceConfig)
+                      (implicit val system: ActorSystem, val materializer: Materializer, protected val executionContext: ExecutionContext)
+  extends WorkspaceSupport with AttributeSupport with LibraryPermissionsSupport with Retry with LazyLogging {
+
+  import dataSource.dataAccess.driver.api._
+
+  /**
+    * Public method to be used for creating a new Workspace.  The bulk of the creation process is performed by
+    * `createWorkspaceInternal`
+    *
+    * @param workspaceRequest
+    * @param span
+    * @return
+    */
+  def createWorkspace(workspaceRequest: WorkspaceRequest, span: Span = null): Future[Workspace] = {
+    validateAttributeNamespace(workspaceRequest.attributes.keys)
+    createWorkspaceInternal(workspaceRequest, span)
+  }
+
+  // TODO: What's up with CloneWorkspace in Orch? Need to make sure this is doing same things as Orch
+  /**
+    * Public method for cloning an existing Workspace.  The creation of the new cloned Workspace will be performed by
+    * `createWorkspaceInternal`, but this method will need to do some work before that to make sure the user has the
+    * necessary permissions on the source Workspace and that any Auth Domain restrictions on the source Workspace are
+    * still in place on the cloned Workspace.  After the cloned Workspace is created, this method will also do things
+    * like copying Entities, Method Configs, and bucket data from the source Workspace into the cloned Workspace.
+    *
+    * @param sourceWorkspaceName
+    * @param destWorkspaceRequest
+    * @return
+    */
+  def cloneWorkspace(sourceWorkspaceName: WorkspaceName, destWorkspaceRequest: WorkspaceRequest): Future[Workspace] = {
+    destWorkspaceRequest.copyFilesWithPrefix.foreach(prefix => validateFileCopyPrefix(prefix))
+    val (libraryAttributeNames, workspaceAttributeNames) = destWorkspaceRequest.attributes.keys.partition(name => name.namespace == AttributeName.libraryNamespace)
+
+    validateAttributeNamespace(workspaceAttributeNames)
+    validateLibraryAttributeNamespaces(libraryAttributeNames)
+
+    for {
+      sourceWorkspace <- getWorkspaceIfHasUserHasAction(sourceWorkspaceName, SamWorkspaceActions.read)
+      destAuthDomains <- validateAuthDomainsForDestWorkspace(sourceWorkspace.workspaceId, destWorkspaceRequest.authorizationDomain.getOrElse(Set.empty))
+      newAttrs = sourceWorkspace.attributes ++ destWorkspaceRequest.attributes
+      cloneWorkspaceRequest = destWorkspaceRequest.copy(authorizationDomain = Option(destAuthDomains), attributes = newAttrs)
+      destWorkspace <- createWorkspaceInternal(cloneWorkspaceRequest)
+      _ <- copyEntitiesAndMethodConfigs(sourceWorkspace, destWorkspace)
```

Comment:
An inadvertent change here is that creating the workspace and copying the source contents are separate transactions. The copy could fail leaving a created but empty workspace.

Reply:
Yeah, I am open to suggestions here.

The original goal of this refactor was to isolate the db transactions from external API calls, and one of my primary concerns about the resulting code is that by doing this, we have changed the behavior.  So what is acceptable here and what is not?

I also tried to preserve the idea that `cloneWorkspace` does the same thing as `createWorkspace`, plus the couple of extra steps before/after as needed.  To achieve what you're looking for here, avoiding the partially cloned workspace, I think Stairway would be the ideal solution.  Short of that, I think this will take some deeper thought or some finer interweaving of `create` and `clone` maybe.  I will take a look at this now to see if there's an elegant solution.